First pass at a material-style copy/paste toolbar.

mpcomplete opened the pull request:
This mimics the toolbar you see when selecting text in an Android
material app. There's still more to do (like integration with the system
clipboard), but this seemed like a good stopping point.

mpcomplete committed:
First pass at a material-style copy/paste toolbar.

This mimics the toolbar you see when selecting text in an Android
material app. There's still more to do (like integration with the system
clipboard), but this seemed like a good stopping point.

mpcomplete commented on packages/flutter/lib/src/material/text_selection_material.dart:
```diff
@@ -0,0 +1,191 @@
+
+/// Draws a single text selection handle. The [type] determines where the handle
+/// points (e.g. the [left] handle points up and to the right).
+class _TextSelectionHandlePainter extends CustomPainter {
```
This class is unchanged. Just moved from input.dart.

mpcomplete commented on packages/flutter/lib/src/material/text_selection_material.dart:
```diff
@@ -0,0 +1,191 @@
+}
+
+/// Builder for material-style text selection handles.
+Widget textSelectionHandleBuilder(
```
Unchanged, just moved from input.dart.

abarth replied:
I'd rename this to use a verb (e.g., `buildTextSelectionHandle`)

abarth commented on packages/flutter/lib/src/material/text_selection_material.dart:
```diff
@@ -0,0 +1,191 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
```
No need for the `_material` suffix.  The file is in the `material` directory.  We have lots of duplicated file names in different directories.

mpcomplete committed:
rename.file

abarth commented on packages/flutter/lib/src/material/text_selection_material.dart:
```diff
@@ -0,0 +1,191 @@
+      maxWidth: constraints.maxWidth,
+      minHeight: 0.0,
+      maxHeight: constraints.maxHeight
+    );
```
`constraints.loosen()`

abarth commented on packages/flutter/lib/src/material/text_selection_material.dart:
```diff
@@ -0,0 +1,191 @@
+}
+
+/// Builder for material-style copy/paste text selection toolbar.
+Widget textSelectionToolbarBuilder(
```
We usually name functions with verbs, so `buildTextSelectionToolbar`

abarth commented on packages/flutter/lib/src/material/text_selection_material.dart:
```diff
@@ -0,0 +1,191 @@
+    BuildContext context, Point position, TextSelectionDelegate delegate) {
+  final Size screenSize = MediaQuery.of(context).size;
+  return new ConstrainedBox(
+    constraints: new BoxConstraints(maxWidth: screenSize.width, maxHeight: screenSize.height),
```
`new BoxConstraints.loose(screenSize)`

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -181,8 +182,14 @@ class RawInputLine extends Scrollable {
 
+  /// Optional builder function for a set of controls for working with a
+  /// text selection (e.g. copy and paste).
+  final TextSelectionToolbarBuilder selectionToolbarBuilder;
```
These I'd keep as noun phrases because they're variables.

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -24,71 +25,154 @@ enum TextSelectionHandleType { left, right, collapsed }
+  void set inputValue(InputValue value);
+
+  /// The copy/paste buffer. Application-wide.
+  String get pasteBuffer;
```
This should return a `Future<String>` because  reading the platform clipboard might be asynchronous.

mpcomplete replied:
I'm working on the clipboard now, and I suspect there may be bigger changes here. Can I leave this for the next patch?

abarth replied:
Sure.

collinjackson committed:
Support for putting assets in android/assets folder (like ios/Resources)

mpcomplete committed:
abarth

Hixie committed:
Provide framework-only timing information (#3699)

This makes it easier to generate charts that show the engine time and
the framework time separately.

Hixie committed:
Track the worst frame in perf tests (#3642)

* Track the worst frame in perf tests

* Use backticks in "/// Returns `null` if"

collinjackson committed:
Merge pull request #3701 from collinjackson/assets

Support for putting assets in android/assets folder (like ios/Resources)

Hixie committed:
Update engine.version (#3704)

abarth committed:
PathProvider.getTemporaryDirectory() never resolves (#3705)

We were looking for this service in the wrong service provider.

Fixes #3683

Hixie committed:
Fix framework-side time in profile data (#3706)

Hixie committed:
ClipPath (#3695)

* ClipPath

* Add a test for ClipOval and ClipPath

abarth committed:
Add more dartdoc to material.dart (#3702)

This patch completes the circuit on this library. The next step is to go
through and add docs to anything we missed.

devoncarew committed:
make the usage text more visible (#3693)

* make the usage text more visible

* update the look of the usage text

yjbanov committed:
[tools] delete profile_startup.dart (#3703)

The same can now be achieved using `flutter run --trace-startup --no-checked`.

vlidholt committed:
Refactor sprite physics, part 1 (#3711)

Commit:
Gallery demo start-time performance test (#3655)

*  Gallery demo start-time performance test

abarth committed:
Add missing dartdocs from material.dart (#3709)

Making progress towards document all public APIs in material.dart. We're still
missing a few odds and ends (as well as missing docs in tabs.dart,
tooltip.dart, and two_level_list.dart).

abarth commented on packages/flutter/lib/src/material/text_selection.dart:
```diff
@@ -0,0 +1,186 @@
+
+    return new Material(
+      elevation: 1,
+      child: new Container(
```
Do you want to use a ButtonBar here?  That will help get the right padding.

mpcomplete replied:
Hmm.. when I try to use a ButtonBar, there is too much padding, and the `more` menu is placed too far down vertically. With Row, there is less padding than there used to be, but it is still too much.

abarth commented:
![LGTM](http://junkyard.damowmow.com/563)

Have you tested this when you rotate the screen while the menu is open?

abarth commented:
It might be worth adding a test that, for example, select all works.

Hixie committed:
Save some memory by only storing one copy of the animation object... (#3724)

Turns out we were storing the same object in the base class and the
subclass. For pretty much every AnimatedWidget subclass.

pq committed:
Literals get type annotations.

As per the recent fix to the `always_specify_types` lint (https://github.com/dart-lang/linter/issues/199), literal maps and lists are now expected to be explicitly typed.

Running that lint on the repo identifies quite a few spots to update.  This focuses on `flutter_driver` and `flutter_sprites` (somewhat arbitrarily) but the changes are fairly representative.

Note there are a number of places where I made a quick judgement on how specific to make the types.  Feedback on those is welcome.  (Especially as we move forward with more.)

pq committed:
Merge pull request #3727 from pq/annotate_literals

Literals get type annotations.

devoncarew committed:
fix the benchmark file written by the analyze command (#3725)

krisgiesing committed:
Add analysis options near the root of the repository (#3700)

This aligns the analysis options used by 'flutter analyze' with the
options used by Atom.

mpcomplete committed:
merget

mpcomplete committed:
select/copy/paste

mpcomplete commented:
Added a test for select/copy/paste. Will commit once it clears travis.

(Also, looks OK when you rotate the screen with a selection).

mpcomplete committed:
fix lint
